src/tests/ftest/util/container_rf_test_base.py

```diff
@@ -114,76 +114,92 @@ def execute_during_rebuild_cont_rf(self, rf, expect_cont_status="HEALTHY"):
+                self.fail("#Container redundancy factor with an invallid "
+                           "object_class traffic passed, expecting Fail")
+            except DaosTestError as error:
+                self.log.info(error)
+                if der_inval in str(error):
+                    self.log.info("==Negative test, expecting DaosApiError "
+                                  "RC: -1003 found.")
+                else:
+                    self.fail("#Negative test, container redundancy factor "
+                              "test failed, return error RC: -1003 not found")
+
     def execute_rebuild_test(self, create_container=True):
         """Execute the rebuild test steps for container rf test.
 
         Args:
             create_container (bool, optional): should the test create a
                 container. Defaults to True.
         """
-        # Get the test params
+
+        # Get the test params and var
         self.setup_test_pool()
         self.daos_cmd = DaosCommand(self.bin)
         if create_container:
             self.setup_test_container()
+        oclass = self.inputs.object_class.value
         num_of_ranks = len(self.inputs.rank.value)
         rf = ''.join(self.container.properties.value.split(":"))
-        if num_of_ranks > int(re.search(r"rf([0-9]+)", rf).group(1)):
+        rf_num = int(re.search(r"rf([0-9]+)", rf).group(1))
+        if num_of_ranks > rf_num:
             expect_cont_status = "UNCLEAN"
         else:
             expect_cont_status = "HEALTHY"
+        negative_test = True
+        if "OC_SX" in oclass and rf_num < 1:
+            negative_test = False
+        elif ("OC_RP_2" in oclass and rf_num < 2) or (
+            "OC_RP_3" in oclass and rf_num < 3):
+            negative_test = False
         # Create a pool and verify the pool information before rebuild
         self.create_test_pool()
         # Create a container and write objects
-        self.create_test_container()
-        # Verify the rank to be excluded has at least one object
-        self.verify_rank_has_objects()
-        # Start the rebuild process
-        self.start_rebuild_cont_rf(rf)
-        # Execute the test steps during rebuild
-        self.execute_during_rebuild_cont_rf(rf, expect_cont_status)
-        # Refresh local pool and container
-        self.log.info(
-            "==>(6)Check for pool and container info after rebuild.")
-        self.pool.check_pool_info()
-        self.container.check_container_info()
-        # Verify the excluded rank is no longer used with the objects
-        self.verify_rank_has_no_objects()
-        # Verify the pool information after rebuild
-        self.update_pool_verify()
-        self.execute_pool_verify(" after rebuild")
-        # Verify the container data can still be accessed
-        if expect_cont_status == "HEALTHY":
+        self.create_test_container_and_write_obj(negative_test)
+        if negative_test:
```

Comment:
IMO, this logic is kind of confusing because we have negative_test check in two places; here (181, 183) and in create_test_container_and_write_obj (132, 135). How about we move the else block in 184-205 to the if block after 134? or at least we can add the "Negative test passed" message in 144 and get rid of this if block.

Reply:
Thanks for the comment, we should separate the main-rebuild and sub-tests (steps),
184-205 is the block for main-rebuild test.
132-135 is the sub-tests (steps) method of create_test_container_and_write_obj